include/mbedtls/ssl.h

```diff
@@ -2720,17 +2713,29 @@ int mbedtls_ssl_conf_own_cert( mbedtls_ssl_config *conf,
  *                 hence need not be preserved by the caller for the lifetime
  *                 of the SSL configuration.
  *
+ * \note           While this function may be called multiple times to
+ *                 register multiple PSKs, the number of supported PSKs
+ *                 is implementation-defined. Once the limit is reached,
```

Comment:
I don't think we should use "implementation-defined" here. This is the documentation of a library, not of a standard API that's likely to have multiple implementation, so we should rather say that the limit may change in future versions of the library. Also, we should say what the current limit it. Users should be able to know how the code behaves by reading the documentation rather than having to try things out.

Reply:
I tend to disagree here: In specifying the 3.x API, we _do_ describe some form of 'standard' which, throughout the lifecycle of Mbed TLS 3.X, will see multiple different implementations. As such, I think it makes sense to describe it in a way that focuses on the general behaviour, not the specifics of the current implementation. The latter should IMO only be described in an optional note.

Reply:
I've left the text mostly as-is but changed "implementation-defined" into "version-specific". @mpg Is this any better from your PoV?

Reply:
I think it's much better, but I'm not fully happy yet :)

I understand how we can see the documentation of a given version as a form of standard that all future versions until the next major one will implement, but I don't think that's the best mindset when writing documentation.

1. Standards tend to put emphasis on allowing variations and as a result tend to leave a number of things unspecified. This is friendly to implementors but not to users (and as people who code in C, we should know!). I think we should be nice to our users by avoiding uncertain language when we can.
2. Defining a standard is a more complicated task than documenting a specific API. In addition to the "user's needs vs implementor needs" conflict above, it's quite hard to specify something that hasn't been implemented, used and tested already. For example, I agree that we should leave the description of what happens if multiple keys are provided with the same id for when we actually implement that, but then that means the "standard" is not complete now. I think we can be nice to ourselves by doing the simplest thing that meets the needs of our users: that is, writing documentation rather than some kind of standard.

So, I quite strongly think the correct format for this kind of thing is, in the following order: "Here's the current behaviour (with no uncertain wording). Here's what may change in the future." People are, by definition, using the present version before they use future ones. So the first information they need is how the present version works, let's give that info to them immediately. Worrying about the future should only come second :)

Reply:
@mpg But with this approach, don't we encourage users to adapt their code to unstable behavior? In this case, don't even attempt to configure more than one PSK.

Reply:
> So the first information they need is how the present version works, let's give that info to them immediately. 

No, I think that's not what they need to know first. They need to know what is the _guaranteed behaviour_ until the next major release, because that's what they can rely on.

Reply:
@gilles-peskine-arm, do you want to be the mediator here? :-)

Reply:
One thing I meant to say and forgot: the documentation is versioned together with the code, let's take advantage of that.

> They need to know what is the guaranteed behaviour until the next major release, because that's what they can rely on.

Indeed, but in that instance the current behaviour (you can call this once and it will do what you expect) is indeed behaviour that they can rely on until the next major release. (Getting an error on the second call is indeed not something they can rely on, but that's really such an edge case that it really shouldn't get priority over describing what actually usefully works right now.)

Reply:
If there's some aspect of the behavior that we think is likely to change in a minor version, we should explicitly state that this aspect is unspecified. Possibly list the possible behaviors, if that's known and useful.

I would generally avoid mentioning the current behavior since this invites users into assuming that the current behavior won't change. There may be cases where this is useful, but this should be exceptional: only do it if there's a good reason.

```
$ cat include/mbedtls/*.h |grep -ic currently
47
```

Reply:
Thanks @gilles-peskine-arm for chiming in! However I'm afraid this doesn't close it for me: (1) the general principle you state can be applied in very different ways to the current situation (see blow), and (2) I feel the question about "behaviour" is a bit too abstract - here the "behaviour" we're talking about could also be described as "support (or lack thereof) for a feature".

I'll also note that in the GCM PR, for `update_ad()` you [went with](https://github.com/ARMmbed/mbedtls/pull/4342/files#diff-95ff3f090a2ff3fcc6e3740413f55f20b385aee26c6ff098c478887641713e9dR252) "You may only call this once. This may change in the future." which happens to fit the general pattern I'm suggesting of "[first] This is the current state of affairs. [second] This is what may change." You didn't go with "You may call this function multiple times. Except if you call it more than once, it won't do anything useful with the present version."

Let me try to list some options here:

a. (initial state of this PR) You may register multiple key. The limit is implementation-defined.
b. (current state of this PR) You may register multiple keys. The limit is version-dependent. Currently it's 1.
c. (before this PR) You can only register one key. If you try to register many, the last registered wins.
d. You can only register one key. If you try to register more, the behaviour is unspecified.
e. You can only register one key. If you try to register more, currently you'll get an error, but in the future this may become supported.

My favourite is (e) because we share with the user all the information we can give without restricting ourselves for the future. (d) is a close second, because I find it may leave the reader with a greater feeling of uncertainty, without giving us any more freedom as a compensation. Clearly (c) is problematic and this is why we have this PR. I don't like (b) because it starts by giving a wrong impression before correcting it, but it's a strict improvement of (c) and I could live with it. I think (a) is not a strict improvement of (c) because while it solves he problem with (c) it introduces another: giving the impression that we support a feature that we don't.  (I checked, and it seems widely accepted that "multiple" means at least more that one.) The very first quality of documentation is that it should not be in contradiction with reality. We may strive for other qualities as well, but not at the expense of this one.

Now, I could speak about this all day (full disclosure: I don't know why, but I feel I'm over-investing this discussion), but that's probably not the best use of our time, so I'm going to take a break from this discussion, work on other things, and come back later.

Reply:
I wrote my previous comment without having looked at the specific case in this PR, and I still haven't looked at this specific case beyond the information that can be deduced from the conversation here.

The current state of `mbedtls_gcm_update_ad` is peculiar because its documentation constrains both applications and alternative implementations and the two evolve in different directions. If it hadn't been for alt implementations, I would have stopped at “you may only call this function once”. But then alt implementations would not have needed to support multiple calls. This state of affairs where we put requirements on alt that the built-in code doesn't meet is unfortunate and I really hope we can fix this before the 3.0 release.

Now for the case here, I understand that there is a setting that can be conceptually a set of keys, and currently we only allow this set to have a single element but in the future we want to allow multiple elements.

> a. (initial state of this PR) You may register multiple key. The limit is implementation-defined.

Makes sense to me. I'd want to have some way of knowing what the limit is on my implementation: maybe a `#define`. I expect that attempting to register more keys than supported returns an error `MBEDTLS_ERR_IMPLEMENTATION_LIMIT_REACHED` or something.

> b. (current state of this PR) You may register multiple keys. The limit is version-dependent. Currently it's 1.

This could make sense if we were implementing some generic specification for a use case where multiple keys are not needed. But here we control both the specification and the implementation. Specifying multiple keys and implementing only one is contradictory.

> c. (before this PR) You can only register one key. If you try to register many, the last registered wins.

This behavior makes sense. The registered key (singular) is a setting, and you can change the value of this setting again and again if you want. It might not be the most useful behavior if the registered keys are conceptually a set, but it is internally consistent, clear and unsurprising.

> d. You can only register one key. If you try to register more, the behaviour is unspecified.

This would make sense if we didn't know at this time whether we want to support registering multiple keys.

> e. You can only register one key. If you try to register more, currently you'll get an error, but in the future this may become supported.

This makes sense if we know we want to support registering multiple keys in the future.

Given what I've read in this discussion, I favor (e). I'd be equally fine with (c) from the point of view of internal consistency, but clearly it's not the semantics we want. (a) is a lie. (b) is logically equivalent to (e) but explained in a confusing way. (d) is acceptable but if we can already guarantee that calling the function won't override a previous key, that's better.

Reply:
I retain my reservation that the focus on the current limitation of a single key encourages people to adjust their code to this limit, instead of writing the code in a portable way for 3.X.

Reply:
However, the discussion seems to be growing out of proportion, and I feel we're all too busy to dwell on this for too long. In the interest of moving forward, I'll go with a wording in the spirit of (e).